src/config.rs

```diff
@@ -21,58 +123,59 @@ struct JsonEntry {
-    /// into the ConfigStore and is more performant normally than calling get_setting manually for each
-    /// setting.
-    fn get_all_settings(&self) -> Result<HashMap<String, Setting>>;
-}
-
-/// Configuration store is the place where the gosub engine can find all configurable options
+/// Configuration storage is the place where the gosub engine can find all configurable options
 pub struct ConfigStore {
     /// A hashmap of all settings so we can search o(1) time
-    settings: HashMap<String, Setting>,
+    /// The mutex allows to share between multiple threads,
+    /// The refcell allows us to use mutable references in a non-mutable way (ie: settings can be
+    /// stored while doing a immutable get())
+    settings: std::sync::Mutex<std::cell::RefCell<HashMap<String, Setting>>>,
     /// A hashmap of all setting descriptions, default values and type information
     settings_info: HashMap<String, SettingInfo>,
     /// Keys of all settings so we can iterate keys easily
     setting_keys: Vec<String>,
     /// The storage adapter used for persisting and loading keys
-    storage: Box<dyn Store>,
+    storage: Box<dyn StorageAdapter>,
 }
 
-impl ConfigStore {
-    /// Creates a new store with the given storage adapter and preloads the store if needed
-    pub fn from_storage(storage: Box<dyn Store>, preload: bool) -> Result<Self> {
+impl Default for ConfigStore {
+    fn default() -> Self {
         let mut store = ConfigStore {
-            settings: HashMap::new(),
+            settings: std::sync::Mutex::new(std::cell::RefCell::new(HashMap::new())),
             settings_info: HashMap::new(),
             setting_keys: Vec::new(),
-            storage,
+            storage: Box::new(MemoryStorageAdapter::new()),
         };
 
-        // Populate the settings from the json file
-        store.populate_settings()?;
+        // Populate the store with the default settings. They may be overwritten by the storage
+        // as soon as one is added with config::config_store()::set_storage()
+        let _ = store.populate_default_settings();
+        store
+    }
+}
+
+impl ConfigStore {
+    /// Sets a new storage engine and updates all settings in the config store according to what
+    /// is written in the storage. Note that it will overwrite any current settings in the config
+    /// store. Take this into consideration when using this function to switch storage engines.
+    pub fn set_storage(&mut self, storage: Box<dyn StorageAdapter>) {
+        self.storage = storage;
 
-        // preload the settings if requested
-        if preload {
-            let all_settings = store.storage.get_all_settings()?;
+        // Find all keys, and add them to the configuration store
+        if let Ok(all_settings) = self.storage.all() {
             for (key, value) in all_settings {
-                store.settings.insert(key, value);
+                self.settings
```

Comment:
lock the mutex,
unwrap the mutex guard,
borrow the refcel as mutable reference
and insert the key/value into the hashmap